<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="cart_pole">

<!-- pi -->
<xacro:include filename="$(find cart_pole)/urdf/robot.gazebo" />

<!-- Launch arguments which will be passed to xacro file when /start_simulation service is called -->
<xacro:arg name="mc" default="1.5"/>
<xacro:arg name="mp" default="0.5"/>
<xacro:arg name="l"  default="1.0"/>
<xacro:arg name="k1" default="-0.8935"/>
<xacro:arg name="k2" default="16.7650"/>
<xacro:arg name="k3"  default="-1.7750"/>
<xacro:arg name="k4" default="5.0945"/>


<xacro:property name="mc" value="${$(arg mc) - 0.5}" />
<xacro:property name="mp" value="${$(arg mp)}" />
<xacro:property name="l" value="${$(arg l)}" />


<!-- dummy_link -->
<link name="dummy_link" />

<!-- dummy_joint -->
<joint name="dummy_joint" type="fixed">
    <parent link="dummy_link" />
    <child link="base" />
</joint>

<!-- base link -->
<!-- cart or mass m_c in dynamic model, particle mass with no inertia -->
<link name="base">
    <visual>
        <origin xyz="0 0 0.2" />
        <geometry>
            <box size="0.05 0.618 0.05" />
        </geometry>
    </visual>
    <collision>
        <origin xyz="0 0 0.2" />
        <geometry>
            <box size="0.05 0.618 0.05" />
        </geometry>
    </collision>
    <inertial>
        <origin xyz="0 0 0.2" />
        <mass value="${mc}" />
        <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0" />
    </inertial>
</link>

<!-- right_wheel -->
<!-- has I_w = izz inertia -->
<link name="right_wheel">
    <visual>
        <geometry>
            <cylinder radius="0.2" length="0.04" />
        </geometry>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.2" length="0.04" />
        </geometry>
    </collision>
    <inertial>
        <mass value="0.25" />
        <inertia ixx="0.0006333" ixy="0.0" ixz="0.0" iyy="0.0006333" iyz="0.0" izz="0.00125" />
    </inertial>
</link>

<!-- right_wheel_joint -->
<!-- has d_1 damping ratio -->
<joint name="right_wheel_joint" type="continuous">
    <parent link="base" />
    <child link="right_wheel" />
    <origin xyz="0.0 -0.329 0.2" rpy="${pi/2} 0.0 0.0" />
    <axis xyz="0 0 1" />
    <limit effort="10.0" velocity="100.0" />
    <dynamics damping="0.01" />
</joint>

<!-- left_wheel -->
<!-- has I_w = izz inertia -->
<link name="left_wheel">
    <visual>
        <geometry>
            <cylinder radius="0.2" length="0.04" />
        </geometry>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.2" length="0.04" />
        </geometry>
    </collision>
    <inertial>
        <mass value="0.25" />
        <inertia ixx="0.0006333" ixy="0.0" ixz="0.0" iyy="0.0006333" iyz="0.0" izz="0.00125" />
    </inertial>
</link>

<!-- left_wheel_joint -->
<!-- has d_1 damping ratio -->
<joint name="left_wheel_joint" type="continuous">
    <parent link="base" />
    <child link="left_wheel" />
    <origin xyz="0.0 0.329 0.2" rpy="${pi/2} 0.0 0.0" />
    <axis xyz="0 0 1" />
    <limit effort="10.0" velocity="100.0" />
    <dynamics damping="0.01" />
</joint>

<!-- pendulum link -->
<!-- this link is mass less just like the configuration figure -->
<link name="pendulum">
    <visual>
        <geometry>
            <box size="0.05 ${l} 0.05" />
        </geometry>
    </visual>
    <collision>
        <geometry>
            <box size="0.05 ${l} 0.05" />
        </geometry>
    </collision>
</link>

<!-- pendulum_joint -->
<joint name="pendulum_joint" type="fixed">
    <parent link="base" />
    <child link="pendulum" />
    <origin xyz="0.0 0.0 ${l/2 + 0.2}" rpy="${pi/2} 0.0 0.0" />
</joint>

<!-- pole link -->
<!-- think of this as mass M_p, a particle mass with no inertia -->
<link name="pole">
    <visual>
        <geometry>
            <sphere radius="0.10" />
        </geometry>
    </visual>
    <collision>
        <geometry>
            <sphere radius="0.10" />
        </geometry>
    </collision>
    <inertial>
        <mass value="${mp}" />
        <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0" />
    </inertial>
</link>

<!-- pole joint -->
<joint name="pole_joint" type="fixed">
    <parent link="base" />
    <child link="pole" />
    <origin xyz="0.0 0.0 ${l + 0.2}" />
</joint>

<!-- LQR controller -->
<!-- gains will be passed to controller via launch arguments -->
<gazebo>
  <plugin name="lqr_controller" filename="liblqr_controller.so">
    <k1>${$(arg k1)}</k1>
    <k2>${$(arg k2)}</k2>
    <k3>${$(arg k3)}</k3>
    <k4>${$(arg k4)}</k4>
  </plugin>
</gazebo>

</robot>